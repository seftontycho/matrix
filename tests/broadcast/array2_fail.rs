#![feature(generic_const_exprs)]

use matrix::prelude::*;

fn main() {
    // impossible shape combinations are:
    // [A, B], [C, B] where A != C and A != 1 and C != 1
    // [A, B], [A, C] where B != C and B != 1 and C != 1
    // [A, B], [C, D] where A != C and B != C and A, B, C, D != 1

    a_b_c_b();
    a_b_a_c();
    a_b_c_d();
}

/// This is case: [A, B], [C, B] where A != C and A != 1 and C != 1
fn a_b_c_b() {
    let a = Array2::from([[1, 2], [3, 4], [5, 6]]);
    let b = Array2::from([[1, 2], [3, 4]]);
    let op: fn(&i32, &i32) -> i32 = |a, b| a * b;

    let c = a.broadcast(b, op);
}

/// This is case: [A, B], [A, C] where B != C and B != 1 and C != 1
fn a_b_a_c() {
    let a = Array2::from([[1, 2], [3, 4]]);
    let b = Array2::from([[1, 2, 3], [4, 5, 6]]);
    let op: fn(&i32, &i32) -> i32 = |a, b| a * b;

    let c = a.broadcast(b, op);
}

/// This is case: [A, B], [C, D] where A != C and B != C and A, B, C, D != 1
fn a_b_c_d() {
    let a = Array2::from([[1, 2, 3], [4, 5, 6]]); // [2, 3]
    let b = Array2::from([[1, 2], [3, 4], [5, 6]]); // [3, 2]
    let op: fn(&i32, &i32) -> i32 = |a, b| a * b;

    let c = a.broadcast(b, op);
}
