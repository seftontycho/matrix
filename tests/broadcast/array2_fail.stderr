error[E0631]: type mismatch in function arguments
  --> tests/broadcast/array2_fail.rs:22:28
   |
22 |     let c = a.broadcast(b, op);
   |               ---------    ^^
   |               |            |
   |               |            expected due to this
   |               |            found signature defined here
   |               required by a bound introduced by this call
   |
   = note: expected function signature `for<'a, 'b> fn(&'a _, &'b matrix::prelude::Array2<{integer}, 2, 2>) -> _`
              found function signature `for<'a, 'b> fn(&'a _, &'b i32) -> _`
note: required by a bound in `broadcast`
  --> src/broadcast.rs
   |
   |     fn broadcast(self, rhs: Rhs, op: impl Fn(&T, &U) -> V) -> Self::Output;
   |                                           ^^^^^^^^^^^^^^^ required by this bound in `Broadcast::broadcast`
help: consider wrapping the function in a closure
   |
22 |     let c = a.broadcast(b, |arg0: &i32, arg1| op(arg0, /* &i32 */));
   |                            ++++++++++++++++++   ++++++++++++++++++

error[E0631]: type mismatch in function arguments
  --> tests/broadcast/array2_fail.rs:31:28
   |
31 |     let c = a.broadcast(b, op);
   |               ---------    ^^
   |               |            |
   |               |            expected due to this
   |               |            found signature defined here
   |               required by a bound introduced by this call
   |
   = note: expected function signature `for<'a, 'b> fn(&'a _, &'b matrix::prelude::Array2<{integer}, 2, 3>) -> _`
              found function signature `for<'a, 'b> fn(&'a _, &'b i32) -> _`
note: required by a bound in `broadcast`
  --> src/broadcast.rs
   |
   |     fn broadcast(self, rhs: Rhs, op: impl Fn(&T, &U) -> V) -> Self::Output;
   |                                           ^^^^^^^^^^^^^^^ required by this bound in `Broadcast::broadcast`
help: consider wrapping the function in a closure
   |
31 |     let c = a.broadcast(b, |arg0: &i32, arg1| op(arg0, /* &i32 */));
   |                            ++++++++++++++++++   ++++++++++++++++++

error[E0631]: type mismatch in function arguments
  --> tests/broadcast/array2_fail.rs:40:28
   |
40 |     let c = a.broadcast(b, op);
   |               ---------    ^^
   |               |            |
   |               |            expected due to this
   |               |            found signature defined here
   |               required by a bound introduced by this call
   |
   = note: expected function signature `for<'a, 'b> fn(&'a _, &'b matrix::prelude::Array2<{integer}, 3, 2>) -> _`
              found function signature `for<'a, 'b> fn(&'a _, &'b i32) -> _`
note: required by a bound in `broadcast`
  --> src/broadcast.rs
   |
   |     fn broadcast(self, rhs: Rhs, op: impl Fn(&T, &U) -> V) -> Self::Output;
   |                                           ^^^^^^^^^^^^^^^ required by this bound in `Broadcast::broadcast`
help: consider wrapping the function in a closure
   |
40 |     let c = a.broadcast(b, |arg0: &i32, arg1| op(arg0, /* &i32 */));
   |                            ++++++++++++++++++   ++++++++++++++++++
